Source/NSGridView.m

```diff
@@ -0,0 +1,1233 @@
+- (NSArray *) _viewsForRowAtIndex: (NSUInteger)rowIndex;
+- (NSArray *) _cellsForColumnAtIndex: (NSUInteger)columnIndex;
+- (NSArray *) _viewsForColumnAtIndex: (NSUInteger)columnIndex;
+@end
+
+@implementation NSGridView (Private)
+- (NSRect) _prototypeFrame
+{
+  NSRect vf = [self frame];
+  NSRect f = NSMakeRect(0.0, 0.0,
+                        (vf.size.width / [self numberOfColumns] - _columnSpacing),
+                        (vf.size.height / [self numberOfRows] - _rowSpacing));
+
+  return f;
+}
+
+- (NSArray *) _cellsForRowAtIndex: (NSUInteger)rowIndex
+{
+  NSMutableArray *result = [NSMutableArray arrayWithCapacity: [_columns count]];
+  NSGridRow *row = [_rows objectAtIndex: rowIndex];
+  
+  FOR_IN(NSGridCell*, c, _cells)
+    {
+      if ([c row] == row)
+        {          
+          [result addObject: c];
+        }
+    }
+  END_FOR_IN(_cells);
+  
+  return result;
+}
+
+- (NSArray *) _viewsForRowAtIndex: (NSUInteger)rowIndex
+{
+  NSArray *cells = [self _cellsForRowAtIndex: rowIndex];
+  NSMutableArray *result = [NSMutableArray arrayWithCapacity: [_columns count]];
+  
+  FOR_IN(NSGridCell*, c, cells)
+    {
+      NSView *v = [c contentView];
+      if (c != nil)
+        {
+          [result addObject: v];
+        }
+      else
+        {
+          [result addObject: [NSGridCell emptyContentView]];
+        }
+    }
+  END_FOR_IN(cells);
+
+  return result;
+}
+
+- (NSArray *) _cellsForColumnAtIndex: (NSUInteger)columnIndex
+{
+  NSMutableArray *result = [NSMutableArray arrayWithCapacity: [_rows count]];
+  NSGridColumn *col = [_columns objectAtIndex: columnIndex];
+
+  FOR_IN(NSGridCell*, c, _cells)
+    {
+      if ([c column] == col)
+        {          
+          [result addObject: c];
+        }
+    }
+  END_FOR_IN(_cells);
+  
+  return result;
+}
+
+- (NSArray *) _viewsForColumnAtIndex: (NSUInteger)columnIndex
+{
+  NSArray *cells = [self _cellsForColumnAtIndex: columnIndex];
+  NSMutableArray *result = [NSMutableArray arrayWithCapacity: [_rows count]];
+  
+  FOR_IN(NSGridCell*, c, cells)
+    {
+      NSView *v = [c contentView];
+      if (c != nil)
+        {
+          [result addObject: v];
+        }
+      else
+        {
+          [result addObject: [NSGridCell emptyContentView]];
+        }
+    }
+  END_FOR_IN(cells);
+
+  return result;
+}
+@end
+
+
+// Private methods...
+@implementation NSGridCell (Private)
+- (void) _setOwningRow: (NSGridRow *)r
+{
+  _owningRow = r;
+}
+
+- (void) _setOwningColumn: (NSGridColumn *)c
+{
+  _owningColumn = c;
+}
+@end
+
+@implementation NSGridView
+
++ (void) initialize
+{
+  if (self == [NSGridView class])
+    {
+      [self setVersion: 1];
+    }
+}
+
+- (void) _refreshCells
+{
+  if ([self isHidden])
+    {
+      return;
+    }
+  else
+    {
+      NSUInteger i = 0;
+      NSUInteger num_col = [self numberOfColumns];
+      NSRect f = [self frame];
+      CGFloat current_x = 0.0, current_y = f.size.height;
+      NSRect p = [self _prototypeFrame];
```

Comment:
The way you compute the prototype frame does not leave any space for padding. This calculation here can never add up.

Reply:
I should add the x padding and y padding in that method so that the calculation is better.

Reply:
No, rather you need to subtract these values before you set the frame of the cell view.

Reply:
Done.